### PR TITLE
handleChange now batches DS.change events

### PR DESCRIPTION
**Pull request by Mitranim**
This solves the problem where a single `DS#inject` call would generate multiple resource-level change notifications.

### CODE REVIEW

**Comment by Mitranim** [comment]
Addition to the pull request: `defineResource` now uses `new Function` instead of `eval`. This fixes the problem where custom classes were sometimes shadowed by native classes with matching names, like `window.Element`.

File: src/datastore/sync_methods/defineResource.js [hunk 1]
```diff
@@ -135,8 +135,7 @@ module.exports = function defineResource (definition) {
     var _class = def['class'] = DSUtils.pascalCase(def.name)
     try {
       if (typeof def.useClass === 'function') {
-        eval(`function ${_class}() { def.useClass.call(this); }`) // eslint-disable-line
-        def[_class] = eval(_class) // eslint-disable-line
+        def[_class] = new Function('def', `return function ${_class}() { def.useClass.call(this); }`)(def) // eslint-disable-line
```
Comment (jmdobry) [inline src/datastore/sync_methods/defineResource.js]: Why this change?

File: src/datastore/sync_methods/defineResource.js [hunk 2]
```diff
@@ -145,8 +144,7 @@ module.exports = function defineResource (definition) {
           return new Ctor()
         })(def.useClass.prototype)
       } else {
-        eval(`function ${_class}() {}`) // eslint-disable-line
-        def[_class] = eval(_class) // eslint-disable-line
+        def[_class] = new Function(`return function ${_class}() {}`)() // eslint-disable-line
```
Comment (jmdobry) [inline src/datastore/sync_methods/defineResource.js]: Why this change?

**Comment by Mitranim** [comment]
Didn't realise it would be included by GitHub into the same PR.

File: src/datastore/sync_methods/inject.js [hunk 3]
```diff
@@ -313,12 +313,10 @@ module.exports = function inject (resourceName, attrs, options) {
     definition.emit('DS.beforeInject', definition, attrs)
   }
 
-  // start the recursive injection of data
+  // start the recursive injection of data; this will also call
+  // `definition.handleChange` for each item
   injected = _inject.call(_this, definition, resource, attrs, options)
 
-  // collection was modified
```
Comment (jmdobry) [inline src/datastore/sync_methods/inject.js]: These lines should not have been removed. Without these lines, `DS.change` will never be emitted on the Resource.

Comment (jmdobry) [inline src/datastore/sync_methods/inject.js]: So it's not actually _never_ but, let me think some more...

Comment (Mitranim) [inline src/datastore/sync_methods/inject.js]: The line I removed was doing essentially the same thing as what the batched notification does now. It would end up being duplicated functionality and data.

Comment (jmdobry) [inline src/datastore/sync_methods/inject.js]: The problem is, this introduces a breaking change for anyone currently relying on the current API. Breaking changes require a major version bump. So, you'll need a way to add the feature without a breaking change, which usually involves making it an option.

Comment (Mitranim) [inline src/datastore/sync_methods/inject.js]: You mean if someone subscribes to `DS.change` on the resource and relies on an individual event emitted for each instance?

Comment (Mitranim) [inline src/datastore/sync_methods/inject.js]: Here's the difference between the old and the new behaviour, as I understand it (hope I got this right).

Old:
* `DS#findAll` -> multiple individual `DS.change` events (one per instance).
* `DS#inject` with an array of attributes -> multiple individual events (one per instance), then a final event with an array of instances.

New:
* `DS#findAll` -> one event with all instances.
* `DS#inject` with an array of attributes -> one event with all instances.

Comment (Mitranim) [inline src/datastore/sync_methods/inject.js]: I'll need to think how to make the improvement without breaking changes. Will probably amend the PR.

**Comment by Mitranim** [comment]
Tried adding this as an option, but was unhappy with the amount of code this introduces (and the mental overhead of another option). I feel like it's better to batch events (if desired) in the application code. Here's what I'm using for React:

```js

export class ReactiveComponent extends React.Component {
  subscriptions = []

  _awaitingUpdate = false

  _queueUpdate = (callback) => {
    if (this._awaitingUpdate) return
    this._awaitingUpdate = true

    setTimeout(() => {
      this._awaitingUpdate = false
      if (typeof callback === 'function') this.forceUpdate(callback)
      else this.forceUpdate()
    }, 0)
  }

  componentWillMount () {
    this.subscriptions.forEach(sub => {
      sub.on('DS.change', this._queueUpdate)
    })
  }

  componentWillUnmount () {
    this.subscriptions.forEach(sub => {
      sub.off('DS.change', this._queueUpdate)
    })
  }
}
```

I think the PR should be cancelled for now. Maybe when someone else raises this topic, they'll have a better idea.

**Comment by Mitranim** [comment]
Closing for now.
